### PR TITLE
Test/unit test for adding a new product endpoint

### PR DESCRIPTION
**Pull request by aidan-marshall**
- Created a unit test to ensure that GetProduct retrieves the right ProductDetailsDto with the correct details
- Added unit test to assert AddNewProductAsync returns a product with the specified details correctly

### CODE REVIEW

File: We.Sell.Bread.API.Unit.Tests/Tests/Services/ProductServiceTests.cs [hunk 1]
```diff
@@ -19,7 +19,46 @@ public void GivenDataExistWhenRetrievingProductReturnTypeMustBeListOfProductDeta
...
+            ProductDetailsDto? product = _productService.GetProduct(new Guid("411fbad3-925e-4044-9269-f962641f9277"));
+            string realName = "Pumpernickel Bread";
+            decimal realPrice = (decimal)20.00;
+            string realDescription = "A dense, dark bread made from coarsely ground whole rye grains";
+            int realStock = 7;
+
+            product.Should().NotBeNull();
+            product.Should().BeOfType(typeof(ProductDetailsDto));
+            Assert.Equal(realName, product.ProductName);
+            Assert.Equal(realPrice, product.Price);
```
Comment (Kwanele664) [inline We.Sell.Bread.API.Unit.Tests/Tests/Services/ProductServiceTests.cs]: Here we have made conscious decision in the project to use fluent assertions as the standard for our automated test cases, as it gives us more functionality than the xunit assertions, see line 34/35. can you adapt your assertions to follow that standard e.g product.Should().Be(realName);

File: We.Sell.Bread.API.Unit.Tests/Tests/Services/ProductServiceTests.cs [hunk 2]
```diff
@@ -19,7 +19,46 @@ public void GivenDataExistWhenRetrievingProductReturnTypeMustBeListOfProductDeta
 
             product.Should().NotBeNull();
             product.Should().BeOfType(typeof(List<ProductDetailsDto>));
-           
+
```
Comment (Kwanele664) [inline We.Sell.Bread.API.Unit.Tests/Tests/Services/ProductServiceTests.cs]: I like your coverage for the positive test cases. However, equally important for testing, is to ensure that your application works as expected under unfavorable conditions (Negative test cases). Using getById methods for example, you also want to validate the user input. For instance, ensuring that the product details will not be returned given an invalid productId.  Please see the customer service tests for possible negative test cases and add them for product service too.

File: We.Sell.Bread.API.Unit.Tests/Tests/Services/ProductServiceTests.cs [hunk 3]
```diff
@@ -19,7 +19,46 @@ public void GivenDataExistWhenRetrievingProductReturnTypeMustBeListOfProductDeta
...
+            Assert.Equal(realDescription, product.Description);
+            Assert.Equal(realStock, product.StockQuantity);
+
+        }
+
+        [Fact]
+        public void AddingANewProductShouldReturnAProductAndShouldContainTheGivenDetails()
+        {
+            string productName = "Ciabatta";
+            decimal price = (decimal)35.00;
```
Comment (Kwanele664) [inline We.Sell.Bread.API.Unit.Tests/Tests/Services/ProductServiceTests.cs]: In c# if you are declaring and assigning a variable inside a method, you do not need to specify the explicit type as the type can be derived from what you have assigned to a variable. To make your code more readable, rather use the generic  "var" type for all your local variables. e.g var productName = "Ciabatta";

File: We.Sell.Bread.API.Unit.Tests/Tests/Services/ProductServiceTests.cs [hunk 4]
```diff
@@ -19,7 +19,46 @@ public void GivenDataExistWhenRetrievingProductReturnTypeMustBeListOfProductDeta
...
             product.Should().BeOfType(typeof(List<ProductDetailsDto>));
-           
+
+        }
+
+        [Fact]
+        public void GivenIdShouldExistAndReturnTypeObjectProductDetailsDto()
+        {
+            ProductDetailsDto? product = _productService.GetProduct(new Guid("411fbad3-925e-4044-9269-f962641f9277"));
+            string realName = "Pumpernickel Bread";
```
Comment (Thumbeza) [inline We.Sell.Bread.API.Unit.Tests/Tests/Services/ProductServiceTests.cs]: Just one thing to add from my side:

As a best practice for writing tests (and code), you do not want to hard code your assignments as this opens up your data to human error and can affect the integrity of our tests. It also means you have to manually create more test data as you add more tests and that gets tedious.

For a more automated and consistent way  of handling test data, make use of libraries that were specifically made for this purpose, unless we need to use real data for our test cases. For this project, we use a library called Faker, that has many convenient methods for creating test data. See the CustomerServiceTests for an example. You can find the list of Faker methods here: https://github.com/oriches/faker-cs?tab=readme-ov-file. To help you a bit, Hacker class has methods for generic verbs, nouns and descriptions.

**Comment by Thumbeza** [comment]
Great effort, please take the time to consider the feedback on the pr and update accordingly. 

Thank you for your contributions. Looking forward to merging this pr :)

**Comment by aidan-marshall** [comment]
Noted. Thanks for the pointers guys, will look to implement them asap

**Comment by aidan-marshall** [comment]
Hey guys updated the pull request let me know what you think 👍

**Review by Kwanele664 — APPROVED** [review]

**Pull request merged**